### ISSUE TITLE
Support output streaming

### ISSUE THREAD

**Issue by UndeadKernel** [dzop/emacs-jupyter#36]
This is more a question rather than an "issue".
One nice feature of using Jupyter Notebooks is that it shows the output of an executing block as it gets processed. That is, in long running blocks of code, printed strings are displayed immediately.
In emacs-jupyter, while running async tasks, print statements are only displayed once code blocks fully complete their execution. I see that there are quite some functions that talk about "result stream". What are these functions meant to do? Is this something that emacs-jupyter could support or is it too far fetched?

**Comment by dzop** [comment]
Are you running code blocks with the `:async yes` header argument? In that case the results are shown as they are received from the kernel like you describe. In the `:async no` case it is more like a regular `org-mode` code block where results are displayed once the code block finishes.

I have the following in my `org-mode` setup so that I don't have to always put `:async yes` on every code block

```emacs-lisp
(with-eval-after-load 'ob-jupyter
    (setf (alist-get :async org-babel-default-header-args:jupyter-python) "yes")
    (setf (alist-get :async org-babel-default-header-args:jupyter-julia) "yes"))
```

Or you can use a property keyword at the top of the file to do it on a per file basis.

```
#+PROPERTY: header-args:jupyter-julia :async yes
```

If you want to look more into how this is done in the code you can start from the `jupyter-org--add-result` function. There you will see that the results are added to a list of results when `:async no` and appended to the buffer when `:async yes`. Then there are a whole bunch of `jupyter-org--append-*` functions and other support functions that are needed to do the appending.

A note about the nomenclature:

- "stream" -> as in a `stdout` stream. 
- "stream result" ->  a chunk of text that the kernel emitted from a stream, e.g. `stdout`.
- "result" -> If "result" appears by itself in a function name without any mention of stream then the function is related to both stream results and value results like images.

**Comment by UndeadKernel** [comment]
It seems that I was testing with an old code-base. Streaming is indeed working as you describe it. Thank you for the nomenclature: I can follow the source code better with it.

I did find an issue which I don't know if it's due to my org configuration: The first end line is deleted and the second line is printed with the first line.
Take this as an example:
```python
#+PROPERTY: header-args:jupyter-python :session py :async yes :exports both :results output

#+BEGIN_SRC jupyter-python
from time import sleep
for i in range(0,4):
    print("foo {}".format(i))
    sleep(1)
#+END_SRC
```
The output is displayed as:
```
#+RESULTS:
#+begin_example
foo 0foo 1
foo 2
foo 3
#+end_example
```
Notice that the first line has both "foo 0" and "foo 1" printed in it.

**Comment by dzop** [comment]
That is interesting because I have written a lot of tests to counteract this problem, but I am seeing this happening still now that you bring it back up. The below line is where the tests for this issue are

https://github.com/dzop/emacs-jupyter/blob/19e18f36f8f6ed0b65a8351de4e62ad5acf6d08d/test/jupyter-test.el#L1394

Note, the `flush=True` will immediately send the print statements to the kernel so that any print messages afterwards come as separate messages.

**Comment by UndeadKernel** [comment]
The test cases make me wonder: When is "#+BEGIN_EXAMPLE" used instead of the colon ( : ) to show results?
In my last example, "foo 0" is written using the colon notation. After printing subsequent "foos", the results block is changed to "#+BEGIN_EXAMPLE". I tried playing around with the variable `org-babel-min-lines-for-block-output` to see if it had any effect.

**Comment by dzop** [comment]
Whenever the number of lines in the output exceeds `org-babel-min-lines-for-block-output` the fixed width block, lines with `:` at the beginning of the line, is converted into an example block. 

But I think the problem is related to the `org-mode` version and how `fixed-width` blocks are handled. In my own setup I am using `org-mode` 9.2 but the tests are running using the built in `org-mode` which is probably why the tests pass. I'll have to look into this further.

Also I don't remember fixed width blocks having a newline at their end whenever I wrote the code. For example when doing

```
print("foo")
```

I get

```
: foo
: 
```

but the code is written to assume that we would get

```
: foo 
```

**Comment by dzop** [comment]
This issue is really about how to recognize that a newline should remain or be removed whenever a new stream result comes in. If we have

```python
print("foo 1")
sleep(1)
print("foo 2")
```

Then we would see two stream messages, the first being `"foo 1\n"` and then 1 second later `"foo 2\n"`. When the second message comes in the buffer looks like

```
#+RESULTS:
: foo 1

...
```

and the second stream message should be appended to give

```
#+RESULTS:
: foo 1
: foo 2

...
```

but now suppose we have the code block

```python
print("foo 1", end="")
sleep(1)
print("foo 2")
```

In this case the stream messages are `"foo 1"` and `"foo2\n"`. The problem here is that the buffer should look identical to the above situation before the second stream message comes in.

```
#+RESULTS:
: foo 1

...
```

It is only when the second stream message gets appended that they should look different

```
#+RESULTS:
: foo 1foo2

...
```

**Comment by UndeadKernel** [comment]
The problem you describe is indeed an interesting one.
I believe we could solve it by tracking whether a line ending is deleted or not.
When appending streamed data, if no line end was deleted, just append the string as is. If a line end was deleted, prepend a line end to the string to add. Then, follow the same logic as before to delete a line end and track the fact.

**Comment by UndeadKernel** [comment]
> Whenever the number of lines in the output exceeds `org-babel-min-lines-for-block-output` the fixed width block, lines with `:` at the beginning of the line, is converted into an example block.
> 
> But I think the problem is related to the `org-mode` version and how `fixed-width` blocks are handled. In my own setup I am using `org-mode` 9.2 but the tests are running using the built in `org-mode` which is probably why the tests pass. I'll have to look into this further.
> 
> Also I don't remember fixed width blocks having a newline at their end whenever I wrote the code. For example when doing
> 
> ```
> print("foo")
> ```
> I get
> 
> ```
> : foo
> : 
> ```
> but the code is written to assume that `org-mode` would remove the final newline and we would get
> 
> ```
> : foo 
> ```

I'm also using Org 9.2 (9.2.1 to be exact). With this version, I see that `org-babel-min-lines-for-block-output` has no effect.

**Comment by dzop** [comment]
Yes we do keep track of the line ending as a text property. That's what the function
 `jupyter-org--mark-stream-result-newline ` does. It adds a `jupyter-stream-newline` text property the last newline of a stream result inside the `org-mode` buffer. Whenever a new stream result is appended, that text property is checked and the newline is kept if it is non-nil.

So this is definitely related to how Org 9.2.1 handles fixed-width blocks with a newline on the last line. In Org 9.2.1, for a stream message of `"foo\n"` I am getting

```
: foo
:
```

In Org 9.1.9

```
: foo
```

I was able to fix the issue by stripping the white-space at the end of the stream result in the Org 9.2.1 case so it gives the same result as in Org 9.1.9 which is much cleaner. I was able to get the tests to pass with Org 9.2.1 so I will be pushing some fixes soon.
